### ISSUE TITLE
[JENKINS-11256] Maven3 parallel build fails with java.util.ConcurrentModificationException in Jenkins

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#11252]
<p>linux, jenkins 1.433, maven 3.0.3, maven project with maven 3 and -T 8 running on a remote ssh slave<br/>
build fails with</p>

<p>&#8212; maven-install-plugin:2.3.1:install (default-install) @ xxxxxxxx &#8212;<br/>
Waiting for Jenkins to finish collecting data<br/>
mavenExecutionResult exceptions not empty<br/>
message : Internal error: java.util.ConcurrentModificationException<br/>
cause : null<br/>
Stack trace : <br/>
org.apache.maven.InternalErrorException: Internal error: java.util.ConcurrentModificationException<br/>
	at org.apache.maven.lifecycle.internal.BuilderCommon.handleBuildError(BuilderCommon.java:128)<span class="error">&#91;INFO&#93;</span> Installing xxxxxxxx.jar to xxxxxxxx</p>

<p>	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:95)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:167)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:164)<br/>
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:269)<br/>
	at java.util.concurrent.FutureTask.run(FutureTask.java:123)<span class="error">&#91;INFO&#93;</span> Compiling 1 source file to xxxxxxxx</p>

<p>	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:651)<br/>
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:676)<br/>
	at java.lang.Thread.run(Thread.java:595)<br/>
Caused by: java.util.ConcurrentModificationException<br/>
	at java.util.ArrayList.writeObject(ArrayList.java:572)<br/>
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at java.util.ArrayList.writeObject(ArrayList.java:569)<br/>
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at java.util.HashMap.writeObject(HashMap.java:1039)<br/>
	at sun.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at hudson.remoting.UserRequest._serialize(UserRequest.java:155)<br/>
	at hudson.remoting.UserRequest.serialize(UserRequest.java:164)<br/>
	at hudson.remoting.UserRequest.<init>(UserRequest.java:62)<br/>
	at hudson.remoting.Channel.callAsync(Channel.java:708)<br/>
	at hudson.maven.Maven3Builder$FilterImpl.executeAsync(Maven3Builder.java:229)<br/>
	at hudson.maven.reporters.MavenFingerprinter.postBuild(MavenFingerprinter.java:96)<br/>
	at hudson.maven.Maven3Builder$MavenExecutionListener.recordProjectSucceeded(Maven3Builder.java:431)<br/>
	at hudson.maven.Maven3Builder$MavenExecutionListener.projectSucceeded(Maven3Builder.java:405)<br/>
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:74)<br/>
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:42)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:91)<br/>
	... 7 more<br/>
<span class="error">&#91;INFO&#93;</span> <br/>
<span class="error">&#91;INFO&#93;</span> &#8212; maven-compiler-plugin:2.3.2:testCompile (default-testCompile) @ xxxxxxxx &#8212;<br/>
channel stopped<br/>
Archiviere Artefakte<br/>
Finished: FAILURE</p>

---
<details><summary><i>Originally reported by <a title="kisa" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kisa">kisa</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-11256" target="_blank">Maven3 parallel build fails with java.util.ConcurrentModificationException in Jenkins</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="kutzi" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>
<li><b>reported by</b>: kisa
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-11-18T14:20:29+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-07
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Does it happen frequently or only once?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=paux">paux</a>:</i>
<p>All three collections in Maven Fingerprinter need to be synchronized (Collections.synchronized...()) schould be sufficient. This problem could occur at several places.</p>

<p>Also, check wether a given file has alread been digested should be converted into a onestep checkAndSet operation.</p>

<p>I will create a pull request for this.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Thanks for the pull request.<br/>
Yes, please create a separate pull request for each issue.</p>

<p>Also, I wonder if the 3 maps/sets are really the problem, as they're (as far as I understand it, are per Maven module and therefore only written in a single thread). My suspicion was much more regarding the FingerprintAction which is added and modified in a non-threadsafe fashion. Have you tried if your change fixes the problem?</p>

<p>I think I'd rather put a synchronized block inside the BuildCallable in postBuild()</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>According to Olivier this might be a bug in Maven 3.0.3:<br/>
<a href="http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=paux">paux</a>:</i>
<p>Short update:</p>

<p>I could not reproduce the problem using a test installation, I will try another test using the exact plugin configuration of the actual system.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=paux">paux</a>:</i>
<p>Found the culprit (no plugin):</p>

<p><tt>hudson.maven.reporters.SurefireArchiver.result</tt> is a <tt>TestResult</tt> object containing an ArrayList (suites). This list is updated whenever a module containing tests completes. Since The Async invoke serializes the Maven3Builder and therefore all its reporters (along with the SurefireArchivers). If the fingerprinter runs during the update of the test result, this error occurs. </p>

<p>Solution: Make the field <tt>hudson.tasks.junit.TestResult.suites</tt> a synchronized list (Collections.synchronizedList) </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=paux">paux</a>:</i>
<p>@kutzi: My development environment seems broken. Could you do that change?</p>

<p>Thx</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Ah, forget what I previously wrote here, I missed an important part of your comments.</p>

<p>Did you manage to reproduce it reliably? If so, do you have a test case?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Christoph Kutzinski<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/b39a310c5165234dab6d7acd3ed82bd399d255eb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/b39a310c5165234dab6d7acd3ed82bd399d255eb</a><br/>
Log:<br/>
  Changelog entry for <a href="https://issues.jenkins.io/browse/JENKINS-11256" title="Maven3 parallel build fails with java.util.ConcurrentModificationException in Jenkins" class="issue-link" data-issue-key="JENKINS-11256"><del>JENKINS-11256</del></a>(cherry picked from commit b835fb15497cf31a5069804a9920a747053be3c4)</p>

<p>Conflicts:</p>

<p>	changelog.html</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Christoph Kutzinski<br/>
Path:<br/>
 src/main/java/hudson/maven/reporters/SurefireArchiver.java<br/>
<a href="http://jenkins-ci.org/commit/maven-plugin/00f1af0b4e71f8c691e36a1bb741eeff7bd512ee" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/maven-plugin/00f1af0b4e71f8c691e36a1bb741eeff7bd512ee</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-11256&#93;</span> SurefireArchiver#result doesn't need to be serialized on each remote call. Fixes ConcurrentModificationExceptions in MavenFingerprinter<br/>
Originally-Committed-As: 1476b2ee446f27b0c4ce85c16c5f984ef84cc8a4</p>
